README.md

```diff
@@ -96,6 +96,53 @@ status:
 ...
 ```
 
+## End to end Testing
+
+There is two type of e2e tests launched on CI.
+
+The first one would just apply the yaml files making sure they don't have any
+syntax issues. Pretty simple one, it just basically check the syntax.
+
+The second one would do some proper functional testing, making sure the task
+actually **ran** properly.
+
+The way the functional tests works is that if you have a directory called
+`tests/` inside the task, it would start creating a random `Namespace`, apply
+the task and then every yaml files that you have in that `tests/` directory.
+
+Usually in these other yaml files you would have a yaml file for the
+test resources (`PipelineResource`) and a yaml files for to run the task
+(`TaskRun`).
+
+Sometime you may need to be able to launch some scripts before applying the
+tested task or the other yaml files. Some may pre-setup something on the
+Namesapce or have to do something externally or sometime you may even want to do
+some manipulation of the main `Task`.
+
+For example on the *image builders* tasks like `kaniko` or `jib` we want to
+upload the tasks to a registry to make sure it is actually built properly. To do
+so we [manipulate](kaniko/tests/apply-task-hook.sh) with a python script the
+`Task` (something we don't want for everyone but only for the tests) to add a
+registry as a `Sidecar` and make sure that the `TaskRun` set the parameters to
+upload there. Simple and straightforward no need to upload to an external image
+registry provider having to setup the tokens and deals with the side effects...
+
+There is two different scripts that are checked if present in the `scripts`,
+those scripts actually sourced via the `source` bash script, so you can output
+some environment variables to it that would be applied :
+
+1. **apply-task-hook.sh**: Script to run before applying the task
+2. **apply-taskrun-hook.sh**: Script to run before applying the taskruns or other yaml files.
+
+There is as well some dynamic variables that can be replaced :
+
+* **SERVICE_ACCOUNT**: The service account the user wants to bind to for running
+  the tasks.
```

Comment:
I agree this is confusing and I actually reworked multiple times the imp along the docs and it ended up more confusing.

I have added those because I have started to implement this on our downstream CI which is OpenShift based and due of the restrictive environement of OpenShift we need to setup a serviceaccount for building those tasks. 

As you stated, tha't something confusing for new authors in a pure K8 world, so I may just going to remove this part and find a way downstream to handle this before running the tasks.